openhcl/openhcl_dma_manager/src/lib.rs

```diff
@@ -0,0 +1,453 @@
+
+            let private_pool = self
+                .private_pool
+                .as_mut()
+                .map(SaveRestore::save)
+                .transpose()
+                .map_err(|e| {
+                    SaveError::ChildError("private pool save failed".into(), Box::new(e))
+                })?;
+
+            Ok(OpenhclDmaManagerState {
+                shared_pool,
+                private_pool,
+            })
+        }
+
+        fn restore(&mut self, state: Self::SavedState) -> Result<(), RestoreError> {
+            match (state.shared_pool, self.shared_pool.as_mut()) {
+                (None, None) => {}
+                (Some(_), None) => {
+                    return Err(RestoreError::InvalidSavedState(anyhow::anyhow!(
+                        "saved state for shared pool but no shared pool"
+                    )))
+                }
+                (None, Some(_)) => {
+                    // BUGBUG Is this right?
+                    // It's possible that previously we did not have a shared
+                    // pool, so there may not be any state to restore.
+                }
+                (Some(state), Some(pool)) => {
+                    pool.restore(state).map_err(|e| {
+                        RestoreError::ChildError("shared pool restore failed".into(), Box::new(e))
+                    })?;
+                }
+            }
+
+            match (state.private_pool, self.private_pool.as_mut()) {
+                (None, None) => {}
+                (Some(_), None) => {
+                    return Err(RestoreError::InvalidSavedState(anyhow::anyhow!(
+                        "saved state for private pool but no private pool"
+                    )))
+                }
+                (None, Some(_)) => {
+                    // BUGBUG Is this right?
+                    // It's possible that previously we did not have a private
+                    // pool, so there may not be any state to restore.
+                }
+                (Some(state), Some(pool)) => {
+                    pool.restore(state).map_err(|e| {
+                        RestoreError::ChildError("private pool restore failed".into(), Box::new(e))
+                    })?;
+                }
+            }
+
+            Ok(())
+        }
+    }
+}
+
+/// A global DMA manager that owns various pools of memory for managing
+/// buffers and clients using DMA.
+#[derive(Inspect)]
+pub struct OpenhclDmaManager {
+    /// Page pool with pages that are mapped with shared visibility on CVMs.
+    shared_pool: Option<PagePool>,
+    /// Page pool with pages that are mapped with private visibility on CVMs.
+    private_pool: Option<PagePool>,
+    #[inspect(skip)]
+    inner: Arc<DmaManagerInner>,
+}
+
+/// The required VTL permissions on DMA allocations.
+#[derive(Inspect)]
+pub enum LowerVtlPermissionPolicy {
+    /// The default permissions for allocations. This may be any combination of
+    /// VTL2 only or all VTLs.
+    Default,
```

Comment:
we talked a bit offline, and it seems like the term `Default` might not be quite right here. Something like `Any` might be better to signal a lack of "constraints" on what policy is appropriate